pkg/connector/ble/ble.go

```diff
@@ -146,50 +225,40 @@ func NewConnection(ctx context.Context, vin string) (*Connection, error) {
 	}
 }
 
-func tryToConnect(ctx context.Context, vin string) (*Connection, error) {
+func tryToConnect(ctx context.Context, vin string, target Advertisement) (*Connection, error) {
 	var err error
-	// We don't want concurrent calls to NewConnection that would defeat
-	// the point of reusing the existing BLE device. Note that this is not
-	// an issue on MacOS, but multiple calls to newDevice() on Linux leads to failures.
 	mu.Lock()
 	defer mu.Unlock()
 
-	if device != nil {
-		log.Debug("Reusing existing BLE device")
-	} else {
-		log.Debug("Creating new BLE device")
-		device, err = newDevice()
-		if err != nil {
-			return nil, fmt.Errorf("failed to find a BLE device: %s", err)
-		}
-		ble.SetDefaultDevice(device)
+	if err = initDevice(); err != nil {
+		return nil, err
 	}
 
-	vinBytes := []byte(vin)
-	digest := sha1.Sum(vinBytes)
+	localName := VehicleLocalName(vin)
 
-	localName := fmt.Sprintf("S%02xC", digest[:8])
-	log.Debug("Searching for BLE beacon %s...", localName)
-	canConnect := false
-	filter := func(adv ble.Advertisement) bool {
-		ln := adv.LocalName()
-		if ln != localName {
-			return false
+	if target == nil {
+		target, err = scanVehicleBeacon(ctx, localName)
+		if err != nil {
+			return nil, fmt.Errorf("ble: failed to scan for %s: %s", vin, err)
 		}
-		canConnect = adv.Connectable()
-		return true
 	}
 
-	client, err := ble.Connect(ctx, filter)
-	if err != nil {
-		return nil, fmt.Errorf("failed to find BLE beacon for %s (%s): %s", vin, localName, err)
+	if target.LocalName() != localName {
+		return nil, fmt.Errorf("ble: beacon with unexpected local name: '%s'", target.LocalName())
 	}
 
-	if !canConnect {
+	if !target.Connectable() {
 		return nil, ErrMaxConnectionsExceeded
 	}
 
-	log.Debug("Connecting to BLE beacon %s...", client.Addr())
+	log.Debug("Dialing to %s (%s)...", target.Addr(), localName)
```

Comment:
Is connecting with `target.Addr()` significantly faster than scanning? If so, then it might be worth modifying the API to accept `target.Addr()` instead of `target`. This could allow clients to cache `target.Addr()` to use for future connections.

We'd need to consider being able to recover in cases where `target.Addr()` changes due to a VCSEC replacement, as well as detecting `ErrMaxConnectionsExceeded`. But optmizing for the common case could be useful.

Reply:
Good idea, but If we would count on "longer term" address caching it would make sense to also seperate this part of the `tryToConnect` method to its own function, since it adds back the same pain point this PR is trying to address. Thoughts?

Reply:
I'm fine punting that improvement to a different PR. The reason I proposed it because implementing it as part of this one would help limit API churn.